accepted/exception-handling.md

```diff
@@ -0,0 +1,514 @@
+remain standard compliant, we will eventually have to implement the
+exception handling proposal in Wasmtime.
+
+# Requirements
+[requirements]: #requirements
+
+* **Exception handling should be a zero-cost feature**, i.e. the
+  implementation must not affect the run-time performance (execution
+  speed, memory usage, etc) of programs that do not use the feature.
+
+* **No dependence on `libunwind`.** This library provides a C API for
+  determining and unwinding call chains in ELF programs. Nonetheless,
+  we do not want to bring it into our trusted computing base (TCB),
+  because we have found its implementations to be buggy in the past,
+  and such bugs can easily compromise the integrity of Wasmtime (some
+  past `libunwind` bugs:
+  [#2808](https://github.com/bytecodealliance/wasmtime/issues/2808),
+  [#3256](https://github.com/bytecodealliance/wasmtime/issues/3256),
+  [#7997](https://github.com/bytecodealliance/wasmtime/issues/7997)).
+
+* **Fast unwinding strategy.** We require unwinding to be fast enough
+  to support [the guest
+  profiler](https://docs.wasmtime.dev/examples-profiling-guest.html). Thus,
+  it is important that the chosen implementation strategy does not box
+  ourselves into a corner. Though, we consider it acceptable if the
+  minimal viable product (MVP) of the implementation does not achieve
+  peak performance for unwinding as long as there is a viable path to
+  improving the performance.
+
+* **Compatibility with standard formats.** Regardless of unwinding
+  strategy, Wasmtime must continue to work with system profilers which
+  suport DWARF or frame pointers, e.g. `perf`. Furthermore, in
+  Cranelift `cg_clif` should be extended with capabilities to leverage
+  the unwind info support to emit appropriate DWARF, Windows
+  Structured Exception Handling, and so forth. Though, we believe a
+  MVP capable of recovering the `vmctx` register should suffice for
+  critical use-cases.
+
+# Non-requirements
+[non-requirements]: #non-requirements
+
+* **No support for the legacy exception handling revision.** The
+  legacy revision is being phased out.
+
+* **No support for unwinding across host frames.** Unwinding across
+  host frames would require implementing a full DWARF unwinder or
+  equivalent. It also seems like it would be exceedingly likely to
+  introduce complicated bugs and likely trigger undefined behavior in
+  our Rust code as well.
+
+  Note: this just not preclude throwing and catching exceptions when
+  there is a host frame present in the call stack, as we can catch and
+  rethrow exceptions in trampolines on the boundary.
+
+# Proposal
+[proposal]: #proposal
+
+This section outlines the technical details of the proposal for
+implementing support for exception handling in Wasmtime and Cranelift.
+
+## CLIF Extensions
+[clif-extensions]: #clif-extensions
+
+### `ir::ExceptionTag` and `ir::ExceptionTable`
+
+An `ir::ExceptionTag` is a newtype of an `i32`. These exception tags
```

Comment:
so, for translation from wasm exceptions, do all exceptions translate to the same tag? if not, how do you handle the `catch_all` or `catch_all_ref` instructions?

Reply:
We should be able to reserve a sentinel value for "all".

Reply:
ok, so then cranelift most likely will need to support a wildcard exception tag, this should probably be added to the rfc

Reply:
This shouldn't require anything from Cranelift, the semantics of individual `ir::ExceptionTag`s and which one is the wildcard sentinel can all happen at the Wasmtime level. The fact that a particular `ir::ExceptionTag` is treated as a wildcard by Wasmtime should not have any affect on what Cranelift does.

Reply:
ah, ok. I thought the unwinder would be a cranelift thing.